WIP: FEATURE: Make plugin compatible to cloud hosting services and improve code

gradinarufelix opened the pull request:
PR for issue #2 

**What does it do?**
This plugin currently does not work with Flownative's Beach hosting service as this uses Google Cloud as resource storage and the file replacement doesn't work there. So I switched the whole process from replacing the old file directly to replacing the whole resource of the thumbnail. 

- [x] Removed all var_dumps and development log messages
- [x] Switched composer version to neos/neos:^4.3
- [x] Cleaned up comments
- [ ] Find solution for redirects (see code comment https://github.com/gesagtgetan/GesagtGetan.KrakenOptimizer/pull/3/files#diff-943bd0ff6300e8f07b2b3bfe1944b3e9R201)

Commit:
TASK: Implement GesagtGetan.KrakenOptimizer in project to test new features

Commit:
BUGFIX: Try to find a way round file_name

Commit:
BUGFIX: Try to find a way round file_name

Commit:
BUGFIX: Try to find a way round file_name

Commit:
Bugfixing the filename issue

Commit:
Remove var_dump in controller

Commit:
Tried to resolve problem with not-persisted thumbnails

Commit:
Change return type

Commit:
Switch aspect to slot

Commit:
Remove composer constraint for the meantime

gerhard-boden commented on Classes/Controller/KrakenController.php:
```diff
@@ -56,4 +72,46 @@ public function replaceLocalFileAction()
+        $resourceIdentifier = $krakenIoResult['resourceIdentifier'];
+        $query = $this->entityManager->createQuery('SELECT t FROM Neos\Media\Domain\Model\Thumbnail t WHERE t.resource = :resource');
+        $query->setParameter('resource', $resourceIdentifier);
+        $query->setMaxResults(1);
```
We should take all thumbnails into account that use the resource

gradinarufelix replied:
Have it on my screen already. Thanks for the remind!

gerhard-boden commented on composer.json:
```diff
@@ -3,7 +3,7 @@
     "name": "gesagtgetan/krakenoptimizer",
     "require": {
-        "neos/neos": "^3.0 || ^4.0",
+        "neos/neos": "*",
```
guess this slipped in here for testing purposes, the next major of this plugin will require 4.3 (for the new ``thumbnail persisted`` signal and maybe more)

gradinarufelix replied:
Yes, I need to do this otherwise my composer goes crazy. Please have an eye on that when we are coming to a release-ready version.

gerhard-boden reviewed (commented):
Looks very promising already! Will undertake a thorough once you signal me that this is ready to go. Thanks a lot for all your work so far! 👍

Commit:
Reenabled resource check

Commit:
Add signal trigger for assetResourceReplaced

Commit:
Moved signal to service class in Neos.Media

rolandschuetz commented:
So I tried this in a Flownative Beach Staging setup, but I got the following error:

```
Original Message: Client error: `POST https://api.kraken.io/v1/upload`

400 Bad Request response: {success":false,"message":"No files found in the request"}
```

gradinarufelix commented:
> So I tried this in a Flownative Beach Staging setup, but I got the following error:
> 
> ```
> Original Message: Client error: `POST https://api.kraken.io/v1/upload`
> 
> 400 Bad Request response: {success":false,"message":"No files found in the request"}
> ```

Interesting. I use [this commit](https://github.com/gesagtgetan/GesagtGetan.KrakenOptimizer/pull/3/commits/baac64bc65896faf6e32920f7263daef8ccb2918) in my project in production and all webhooks are delivered well.

rolandschuetz commented:
But I did check out your branch `gmediaat:bugfix/filename-issue`, so that should already be in my package version...

gerhard-boden committed:
BUGFIX: Fix codesniffer style violations

mostly line lenght issues

gerhard-boden committed:
TASK: Adjust context for tests in circle ci

gerhard-boden committed:
BUGFIX: Remove deprecated  methods and classes

gerhard-boden committed:
BUGFIX: Use correct order for arguments

gerhard-boden committed:
TASK: Use wait option as default in requesOptimizedResource method

Also rename settings property

gerhard-boden committed:
TASK: Remove unused systemlogger

gerhard-boden committed:
TASK: Split up code and make it more testable

Outsource Doctrine query statements to service classes and split up code
into more logical methods to improve readability and make it easier
to write tests.

gerhard-boden committed:
TASK: Improve code style and write better comments

gerhard-boden committed:
BUGFIX: Fix command controller by passing necessary resourceIdentifier

gerhard-boden committed:
BUGFIX: Use less confusing variables

gerhard-boden committed:
TASK: Move redirect generation and resource deletion in seperate method

gerhard-boden committed:
BUGFIX: Return early in if live optimization is turned off

Also improve sanity checks

gerhard-boden committed:
FEATURE: Add improved tests

gerhard-boden committed:
BUGFIX: Add ext-json to dependencies

gerhard-boden committed:
TASK: Remove unnecessary FQCNs

gerhard-boden commented:
So guys, there is good news and bad news:

The good news is:
I FINALLY found the time to test and review this. I fixed the tests and improved the test coverage greatly by adding new ones. I fixed various code sniffer style errors and the live optimization seems to be working great BUT (this leads me to the bad news):

The command `kraken:optimize` is broken since the rewrite. At first only some necessary arguments where missing, but in the end it turned out to be a much bigger problem. It's easy to reproduce, just run the command on the Neos demo page after creating a couple of thumbnails and at the end the of the execution there is a doctrine error 
````
An exception occurred while executing 'UPDATE neos_media_domain_model_thumbnail SET resource = ?
WHERE persistence_object_identifier = ?' with params
["c6f5164d-bed6-43fb-b96c-21f96e1dc263",
"1309360a-5c56-4c62-99eb-a6bcb9b124da"]:

SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry
'c6f5164d-bed6-43fb-b96c-21f96e1dc263' for key 'UNIQ_3A163C49BC91F416'

  Type: Doctrine\DBAL\Exception\UniqueConstraintViolationException
  File: Packages/Libraries/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriv
        er.php
  Line: 55

Nested exception:
SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'c6f5164d-bed6-43fb-b96c-21f96e1dc263' for key 'UNIQ_3A163C49BC91F416'

  Type: Doctrine\DBAL\Driver\PDOException
  Code: 23000
  File: Packages/Libraries/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php
  Line: 119

Nested exception:
SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'c6f5164d-bed6-43fb-b96c-21f96e1dc263' for key 'UNIQ_3A163C49BC91F416'

  Type: PDOException
  Code: 23000
  File: Packages/Libraries/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php
  Line: 117
````
(the identifier vary, so I don't think it's related to broken resources).

I even ended up with a couple of broken thumbnails which is even worse than probably unoptimized thumbnails (since there is no resource on the hard drive even more which is not easily fixed by like clearing the caches). 

So in the current state we can not release this. I tried to figure out why this persistence errors occur, but had no luck.

gerhard-boden commented:
@rolandschuetz This sounds like a missing resource file for the thumbnail. Are there more debugging message in the log? We should deferentially not throw an exception if that happens, but skip it and write some error in the log

gradinarufelix commented:
@gerhard-boden I'm currently on it. It seems that we can't have multiple thumbnails with the same resource. Nevermind, I'll just remove that part.

@rolandschuetz This just also happened to me. It seems that when due to this error the resource file gets deleted, the optimizations fails because of the missing resource file. So I guess when we fix this error, we also solve this issue.

rolandschuetz commented:
@gerhard-boden @gradinarufelix Oh, totally missed that there is an open question. When the fix is ready I can test again.

Commit:
Change version constraint to Neos 5.x

Commit:
Adjust tests to match interface method signature

Commit:
Solve problem with non-existent resources and prevent SQL constraint errors

Commit:
Test only Neos 5.x and bump PHP version to 7.3

Commit:
Implement CodeSniffer suggestions

Commit:
Tell Composer to require the local dev vesion of the package

Commit:
Implement CodeSniffer suggestions

Commit:
Adjust path to bootstrap file in AllTests.xml

Commit:
Fix assertion

Commit:
Fix assertions

Commit:
Bugfix in KrakenService.php: corrected order in array_merge

Commit:
TASK: Install PHP GD extension in CircleCI to fix tests

Commit:
Remove unnecessary injection in test

Commit:
Inject ResourceManager in KrakenController

gerhard-boden committed:
FEATURE: Add optimized testing image

gerhard-boden committed:
TEST: Pass valid external URL to test payload

To create a realistic test scenario the implementation can work
with, we have to provide a valid external testing URL.

Creating URLs for localhost is bound to fail because of
the missing webserver configuration on testing environments.
Furthermore, we are not aiming to test resource delivery from the local
web host in this case.

gerhard-boden committed:
TEST: Use obviously broken URL for test

gerhard-boden committed:
TEST: Make codesniffer ignore line
